### PR TITLE
Add `OpenTelemetry::Instrumentation::LogBackend`

### LINKED ISSUE #1: Create an OTelBackend for Logging
The current logging integration directly injects itself into the Log code. In some cases, this may be what is needed. For situations where one has control over the logging backend(s) that are in use, though, it'd be nice to just have a OTelBackend that would attach the log as an event to the current span, if one exists, and NOP if there is no current span. This should be pretty simple to make.

### PR DESCRIPTION
**Pull request by caspiano**
Adds a `Log::Backend` as described in #1

Whenever a `Log::Entry` is emitted to the backend, the message, in addition to the entry's `data` and `context`, is attached to the current span in an event.

```crystal
# It can be required independently
require "opentelemetry-instrumentation/log_backend"
# Or as a default require
require "opentelemetry-instrumentation"

Log.builder.bind("*", Log::Severity::Warn, OpenTelemetry::Instrumentation::LogBackend.new)
```

Closes #1

### CODE REVIEW

File: src/opentelemetry/instrumentation/crystal/log.cr [hunk 1]
```diff
@@ -59,11 +59,13 @@ unless_enabled?("OTEL_CRYSTAL_DISABLE_INSTRUMENTATION_LOG") do
             backend = @backend
             io = IO::Memory.new
             ShortFormat.format(entry, io)
-            span.add_event("Log.#{entry.severity.label}#{!entry.source.empty? ? " - #{entry.source}" : ""}") do |event|
+            span.add_event("Log.#{entry.severity.label}#{" - #{entry.source}" unless entry.source.empty?}") do |event|
               event["message"] = io.rewind.gets_to_end
               if exception = entry.exception
                 event["exception"] = exception.to_s
-                event["backtrace"] = exception.backtrace.join("\n")
+                if backtrace = exception.backtrace?.try(&.join('\n'))
```
Comment (caspiano) [inline src/opentelemetry/instrumentation/crystal/log.cr:66]: There's a `.not_nil!` lurking within `Exception#backtrace`. My spec seemed to trigger it.. so I added a null-safe workaround.

**Comment by wyhaines** [comment]
OpenTelemetry only allows a small, specific set of types for attribute values: String, Bool, Float64, Int64, Int32, or arrays of those types (arrays of mixed types are not allowed). So, the best thing is to probably call `#to_s` on the Log::Metadata::Value before storing it in the attribute. I would try something simple like this:

```crystal
entry.context.each do |key, value|
    event[key.to_s] = value.to_s
end
```

That might be sufficient to provide what is needed.

**Comment by caspiano** [comment]
I reckon there'll be a nice way to get the right values into the event attributes, I'll continue tomorrow.

**Comment by wyhaines** [comment]
Great. I'll keep an eye out for it or any other questions. Thanks!

**Comment by caspiano** [comment]
@wyhaines Got it, I also decided to add `data` attached to the `Log::Entry` to the event as well.

**Comment by wyhaines** [comment]
Oh, I like that.  Using `String.build` is definitely better than what I did. Thank you.

**Pull request merged**